lib/puppet/provider/exec/powershell.rb

```diff
@@ -25,18 +26,60 @@
         }
   EOT
 
+  def self.upgrade_message
+    Puppet.warning <<-UPGRADE
+The current Puppet version is outdated and uses a library that was
+previously necessary on the current Ruby verison to support a colored console.
+
+Unfortunately this library prevents the PowerShell module from using a shared
+PowerShell process to dramatically improve the performance of resource
+application.
+
+To enable these improvements, it is suggested to upgrade to any x64 version of
+Puppet (including 3.x), or to a Puppet version newer than 3.x.
+    UPGRADE
+  end
+
+  def self.win32console_enabled?
+    @enabled ||= defined?(Win32::Console) && Win32::Console.class == Class
+  end
+
+  if self.win32console_enabled?
+    upgrade_message
+  end
+
+  def self.powershell_args
+    ps_args = ['-NoProfile', '-NonInteractive', '-Sta', '-NoLogo', '-ExecutionPolicy', 'Bypass', '-Command']
+    ps_args << '-' if !self.win32console_enabled?
+    ps_args
+  end
+
+  def ps_manager
+    PuppetX::PowerShell::PowerShellManager.instance("#{command(:powershell)} #{self.class.powershell_args.join(' ')}")
+  end
+
   def run(command, check = false)
-    write_script(command) do |native_path|
-      # Ideally, we could keep a handle open on the temp file in this
-      # process (to prevent TOCTOU attacks), and execute powershell
-      # with -File <path>. But powershell complains that it can't open
-      # the file for exclusive access. If we close the handle, then an
-      # attacker could modify the file before we invoke powershell. So
-      # we redirect powershell's stdin to read from the file. Current
-      # versions of Windows use per-user temp directories with strong
-      # permissions, but I'd rather not make (poor) assumptions.
-      return super("cmd.exe /c \"\"#{native_path(command(:powershell))}\" #{args} -Command - < \"#{native_path}\"\"", check)
+    result = ps_manager.execute(command)
+    stdout = result[:stdout]
+    stderr = result[:stderr]
+
+    output = JSON.parse(stdout)
+
+    exit_status = output['exitcode']
+    if exit_status.nil?
+      # TODO: Should we warn user they did not return an exitcode?
+      exit_status = 0
+    end
+
+    unless stderr.nil?
+      stderr.each do |er|
+        er.each { |e| Puppet.debug "STDERR: #{e.chop}" } unless er.empty?
+      end
     end
+
+    # TODO: figure out how to handle json text here if at all
+    stuff = Puppet::Util::Execution::ProcessOutput.new(output['output'].to_s || '', exit_status)
+    return stuff, stuff
```

Comment:
Barring the silly name, following the lead from https://github.com/puppetlabs/puppet/blob/master/lib/puppet/provider/exec.rb#L74-L77